src/components/cc-addon-admin/cc-addon-admin.js

```diff
@@ -29,143 +29,182 @@ export class CcAddonAdmin extends LitElement {
 
-        ${!loadingError ? html`
-          <cc-block-section>
-            <div slot="title">${i18n('cc-addon-admin.heading.name')}</div>
-            <div slot="info"></div>
-            <div class="one-line-form">
-              <cc-input-text
-                label="${i18n('cc-addon-admin.input.name')}"
-                ?skeleton=${this._skeleton}
-                ?disabled=${isFormDisabled}
-                .value=${this._name}
-                @cc-input-text:input=${this._onNameInput}
-                @cc-input-text:requestimplicitsubmit=${this._onNameSubmit}
-              ></cc-input-text>
-              <cc-button primary ?skeleton=${this._skeleton} ?disabled=${isFormDisabled} @cc-button:click=${this._onNameSubmit}>${i18n('cc-addon-admin.update')}</cc-button>
-            </div>
-          </cc-block-section>
-
-          <cc-block-section>
-            <div slot="title">${i18n('cc-addon-admin.heading.tags')}</div>
-            <div slot="info">${i18n('cc-addon-admin.tags-description')}</div>
-            <div class="one-line-form">
-              <cc-input-text
-                label="${i18n('cc-addon-admin.input.tags')}"
-                ?skeleton=${this._skeleton}
-                ?disabled=${isFormDisabled}
-                .tags=${this._tags}
-                placeholder="${i18n('cc-addon-admin.tags-empty')}"
-                @cc-input-text:tags=${this._onTagsInput}
-                @cc-input-text:requestimplicitsubmit=${this._onTagsSubmit}
-              ></cc-input-text>
-              <cc-button primary ?skeleton=${this._skeleton} ?disabled=${isFormDisabled} @cc-button:click=${this._onTagsSubmit}>${i18n('cc-addon-admin.tags-update')}</cc-button>
-            </div>
-          </cc-block-section>
-
-          <cc-block-section>
-            <div slot="title" class="danger">${i18n('cc-addon-admin.danger-zone')}</div>
-            <div slot="info" class="danger-desc">
-              <p>${i18n('cc-addon-admin.delete-disclaimer')}</p>
-              ${shouldShowVmText ? html`<p>${i18n('cc-addon-admin.delete-vm')}</p>` : ''}
-              ${shouldShowBackupsText ? html`<p>${i18n('cc-addon-admin.delete-backups')}</p>` : ''}
-            </div>
-            <div>
-              <cc-button danger ?skeleton=${this._skeleton} ?disabled=${isFormDisabled} @cc-button:click=${this._onDeleteSubmit}>${i18n('cc-addon-admin.delete')}</cc-button>
-            </div>
-          </cc-block-section>
-        ` : ''}
-
-        ${loadingError ? html`
+        ${this.state.type === 'error' ? html`
           <cc-notice intent="warning" message="${i18n('cc-addon-admin.error-loading')}"></cc-notice>
         ` : ''}
+
+        ${this._shouldRenderContent(this.state)
+          ? this._renderContent(this.state)
+          : ''
+        }
       </cc-block>
     `;
   }
 
+  /**
+   * @param {AddonAdminStateLoaded|AddonAdminStateLoading|AddonAdminStateSaving} state
+   * @returns {TemplateResult}
+   * @private
+   */
+  _renderContent (state) {
```

Comment:
note: this feels a bit weird to have a subrender method that takes a state but also rely on `this` to know what to do. It's no big deal and I don't see how we can improve the current code so, just thinking out loud ;-)

Reply:
We agree with you but we don't see how to improve it so we'll leave it like this.